Using datasets for image search blog post (#228)
huggingface/blog: image-search-datasets.md

## Push all the things to the hub!

<img src="https://i.imgflip.com/613c0r.jpg" alt="Push all the things to the hub">

One of the super awesome things about the 🤗 ecosystem is the Hugging Face Hub. We can use the Hub to access models and datasets. It is often used for sharing work with others, but it can also be a useful tool for work in progress. `datasets` recently added a `push_to_hub` method that allows you to push a dataset to the Hub with minimal fuss. This can be really helpful by allowing you to pass around a dataset with all the transforms etc. already done.

For now, we'll push the dataset to the Hub and keep it private initially.

Depending on where you are running the code, you may need to login using the `huggingface-cli login` command.

``` python
huggingface-cli login

> ```

Reviewer: In a notebook, you can run:

```
from huggingface_hub import notebook_login

notebook_login()
```
which is a bit more user friendly compared to the CLI command. In a terminal, you need to run `huggingface-cli login` indeed.

``` python
dataset.push_to_hub('davanstrien/embellishments-sample', private=True)
```

> **Note**: in a [previous version](https://danielvanstrien.xyz/metadata/deployment/huggingface/ethics/huggingface-datasets/faiss/2022/01/13/image_search.html) of this blog post we had to do a few more steps to ensure images were embedded when using `push_to_hub`. Thanks to [this pull request](https://github.com/huggingface/datasets/pull/3685) we no longer need to worry about these extra steps. We just need to make sure `embed_external_files=True` (which is the default behaviour).

### Switching machines

At this point, we've created a dataset and moved it to the Hub. This means it is possible to pick up the work/dataset elsewhere.

In this particular example, having access to a GPU is important. Using the Hub as a way to pass around our data we could start on a laptop
and pick up the work on Google Colab.

If we move to a new machine, we may need to login again. Once we've done this we can load our dataset

``` python
from datasets import load_dataset

dataset = load_dataset("davanstrien/embellishments-sample", use_auth_token=True)
```
## Creating embeddings 🕸 

> We now have a dataset with a bunch of images in it. To begin creating our image search app, we need to embed these images. There are various ways to try and do this, but one possible way is to use the clip models via the `sentence_transformers` library. The [clip model](https://openai.com/blog/clip/) from OpenAI learns a joint representation for both images and text, which is very useful for what we want to do since we want to input text and get back an image.

Reviewer: ```suggestion
We now have a dataset with a bunch of images in it. To begin creating our image search app, we need to embed these images. There are various ways to try and do this, but one possible way is to use the CLIP models via the `sentence_transformers` library. The [CLIP model](https://openai.com/blog/clip/) from OpenAI learns a joint representation for both images and text, which is very useful for what we want to do since we want to input text and get back an image.
```

We can download the model using the `SentenceTransformer` class.

``` python

> from sentence_transformers import SentenceTransformer, util
> 
> model = SentenceTransformer('clip-ViT-B-32')

Reviewer: ```suggestion
from sentence_transformers import SentenceTransformer

model = SentenceTransformer('clip-ViT-B-32')
```
Not sure whether "util" is needed here.

```

> This model will take as input either an image or some text and return an embedding. We can use the `datasets` `map` method to encode all our images using this model. When we call map, we return a dictionary with the key `embeddings` containing the embeddings returned by the embedding model. We also pass `device='cuda'` when we call model; this ensures that we're doing the encoding on the GPU.
> 

Reviewer: ```suggestion
This model will take as input either an image or some text and return an embedding. We can use the `datasets` `map` method to encode all our images using this model. When we call map, we return a dictionary with the key `embeddings` containing the embeddings returned by the model. We also pass `device='cuda'` when we call the model; this ensures that we're doing the encoding on the GPU.

```

``` python
ds_with_embeddings = dataset.map(
    lambda example: {'embeddings':model.encode(example['image'], device='cuda')}, batch_size=32)

> ```

Reviewer: Shouldn't you also set `batched=True`? Since `batched=False` by default.

Author: good catch :)

<h1> Image search with 🤗 datasets </h1> 

<div class="blog-metadata">
    <small>Published March 8, 2022.</small>
    <a target="_blank" class="btn no-underline text-sm mb-5 font-sans" href="https://github.com/huggingface/blog/blob/master/image-search-datasets.md">
        Update on GitHub
    </a>
</div>

<div class="author-card">
    <a href="/davanstrien">
        <img class="avatar avatar-user" src="https://avatars.githubusercontent.com/u/8995957?v=4" title="Gravatar">
        <div class="bfc">
            <code>davanstrien</code>
            <span class="fullname">Daniel van Strien</span>
            <span class="bg-gray-100 rounded px-1 text-gray-600 text-sm font-mono">guest</span>
        </div>
    </a>
</div>

<a target="_blank" href="https://colab.research.google.com/gist/davanstrien/e2c29fbbed20dc767e5a74e210f4237b/hf_blog_image_search.ipynb">
    <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>

🤗 [`datasets`](https://huggingface.co/docs/datasets/) is a library that makes it easy to access and share datasets. It also makes it easy to process data efficiently -- including working with data which doesn't fit into memory.

> When `datasets` was first launched it was more usually associated with text data. However, recently, `datasets` has added increased support for images. In particular, there is now a `datasets` [feature type for images](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=image#datasets.Image). A previous [blog post](https://huggingface.co/blog/fine-tune-vit) showed how `datasets` can be used with 🤗 `transformers` to train an image classification model. In this blog post, we'll see how we can combine `datasets` and a few other libraries to create an image search application.

Reviewer: ```suggestion
When `datasets` was first launched, it was more usually associated with text data. However, recently, `datasets` has added increased support for images. In particular, there is now a `datasets` [feature type for images](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=image#datasets.Image). A previous [blog post](https://huggingface.co/blog/fine-tune-vit) showed how `datasets` can be used with 🤗 `transformers` to train an image classification model. In this blog post, we'll see how we can combine `datasets` and a few other libraries to create an image search application.
```

<span style="color: #000080; text-decoration-color: #000080">╰───────────────────────────────────────────────────────────────────────────────────────────╯</span>
</pre>

We can see there a few different ways in which we can pass in our images. We'll come back to this in a little while.

A really nice feature of the `datasets` library (beyond the functionality for processing data, memory mapping etc.) is that you get
some nice things 'for free'. One of these is the ability to add a [`faiss`](https://github.com/facebookresearch/faiss) index to a dataset. [`faiss`](https://github.com/facebookresearch/faiss) is a ["library for efficient similarity search and clustering of dense
vectors"](https://github.com/facebookresearch/faiss).

> The `datasets` [docs](https://huggingface.co/docs/datasets) show an [example](https://huggingface.co/docs/datasets/faiss_es.html#id1) of using a `faiss` index for text retrieval. In this post we'll see if we can do the same for images.

Reviewer: ```suggestion
The `datasets` [docs](https://huggingface.co/docs/datasets) shows an [example](https://huggingface.co/docs/datasets/faiss_es.html#id1) of using a `faiss` index for text retrieval. In this post we'll see if we can do the same for images.
```

## The dataset: "Digitised Books - Images identified as Embellishments. c. 1510 - c. 1900"

> This is a dataset of images which have been pulled from a collection of digitised books from the British Library. These images come from books across a wide time period and from a broad range of domains. The images were extracted using information contained in the OCR output for each book. As a result, it's known which book the images came from but not necessarily anything else about that image i.e. what it is of.

Reviewer: ```suggestion
This is a dataset of images which have been pulled from a collection of digitised books from the British Library. These images come from books across a wide time period and from a broad range of domains. The images were extracted using information contained in the OCR output for each book. As a result, it's known which book the images came from, but not necessarily anything else about that image i.e. what it is of.
```
That last part is a bit weird to me "what it is of"? What does that mean? :)

## Creating our dataset 

Our dataset consists of a folder containing subdirectories inside which are images. This is a fairly standard format for sharing image datasets. Thanks to a recently merged [pull request](https://github.com/huggingface/datasets/pull/2830) we can directly load this dataset using `datasets` `ImageFolder` loader 🤯

```python
from datasets import load_dataset
dataset = load_dataset("imagefolder", data_files="https://zenodo.org/record/6224034/files/embellishments_sample.zip?download=1")
```

Let's see what we get back.

```python
dataset
```

```
DatasetDict({
    train: Dataset({
        features: ['image', 'label'],
        num_rows: 10000
    })
})
```

> We can get back a DatasetsDict, and we have a Dataset with image and label features. Since we don't have any train/validation splits here, let's grab the train part of our dataset. Let's also take a look at one example from our dataset to see what this looks like.

Reviewer: ```suggestion
We can get back a `DatasetDict`, and we have a Dataset with image and label features. Since we don't have any train/validation splits here, let's grab the train part of our dataset. Let's also take a look at one example from our dataset to see what this looks like.
```

```python

> dataset[0]

Reviewer 2: I think one line is missing here:
```suggestion
dataset = dataset["train"]
dataset[0]
```

``` python
ds_with_embeddings['train'].add_faiss_index(column='embeddings')
```

``` 
Dataset({
        features: ['fname', 'year', 'path', 'image', 'embeddings'],
        num_rows: 10000
    })
```

## Image search

> **Note** that these examples were generated from the full version of the dataset so you may get slightly different results.

We now have everything we need to create a simple image search. We can use the same model we used to encode our images to encode some input text. This will act as the prompt we try and find close examples for. Let's start with 'a steam engine'.

``` python
prompt = model.encode("A steam engine")
```

We can use another method from the datasets library `get_nearest_examples` to get images which have an embedding close to our input prompt embedding. We can pass in a number of results we want to get back.

``` python

> scores, retrieved_examples = ds_with_embeddings['train'].get_nearest_examples('embeddings', prompt,k=9)

Reviewer: ```suggestion
scores, retrieved_examples = ds_with_embeddings['train'].get_nearest_examples('embeddings', prompt, k=9)
```

```

We can index into the first example this retrieves:

``` python
retrieved_examples['image'][0]
```

<img src="assets/53_image_search_datasets/search_result.jpg" alt="An image of a factory">

This isn't quite a steam engine, but it's also not a completely weird result. We can plot the other results to see what was returned.

``` python
import matplotlib.pyplot as plt
```

``` python
plt.figure(figsize=(20, 20))
columns = 3
for i in range(9):
    image = retrieved_examples['image'][i]
    plt.subplot(9 / columns + 1, columns, i + 1)
    plt.imshow(image)
```

<img src="assets/53_image_search_datasets/steam_engine_search_results.jpg">

> Some of these results look fairly close to our input prompt. We can wrap
> this in a function so can more easily play around with different prompts

Reviewer: ```suggestion
Some of these results look fairly close to our input prompt. We can wrap
this in a function so we can more easily play around with different prompts
```

``` python
def get_image_from_text(text_prompt, number_to_retrieve=9):
    prompt = model.encode(text_prompt)

>     scores, retrieved_examples = ds_with_embeddings['train'].get_nearest_examples('embeddings', prompt,k=number_to_retrieve)

Reviewer: ```suggestion
    scores, retrieved_examples = ds_with_embeddings['train'].get_nearest_examples('embeddings', prompt, k=number_to_retrieve)
```

    plt.figure(figsize=(20, 20))
    columns = 3
    for i in range(9):
        image = retrieved_examples['image'][i]
        plt.title(text_prompt)
        plt.subplot(9 / columns + 1, columns, i + 1)
        plt.imshow(image)
```

``` python
get_image_from_text("An illustration of the sun behind a mountain")
```
<img src="assets/53_image_search_datasets/sun_behind_mountain.jpg">

### Trying a bunch of prompts ✨

> Now we have a function for getting a few results we can try a bunch of

Reviewer: ```suggestion
Now we have a function for getting a few results, we can try a bunch of
```

```

``` python
for prompt in prompts:
    get_image_from_text(prompt)
```

<img src="assets/53_image_search_datasets/musical_instrument.jpg">

<img src="assets/53_image_search_datasets/guitar.jpg">

<img src="assets/53_image_search_datasets/an_animal.jpg">

<img src="assets/53_image_search_datasets/cat_or_dog.jpg">

<img src="assets/53_image_search_datasets/an_empty_abyss.jpg">

We can see these results aren't always right but they are usually some reasonable results in there. It already seems like this could be useful for searching for a the semantic content of an image in this dataset. However we might hold off on sharing this as is...

## Creating a Hugging Face Space? 🤷🏼 

One obvious next step for this kind of project is to create a Hugging Face [Spaces](https://huggingface.co/spaces/launch) demo. This is what I've done for other [models](https://huggingface.co/spaces/BritishLibraryLabs/British-Library-books-genre-classifier-v2)

It was a fairly simple process to get a [Gradio app setup](https://gradio.app/) from the point we got to here. Here is a screenshot of this app:

> <img src="assets/53_image_search_datasets/spaces_image_search.jpg" alt="Screnshot of Gradio search app">

Reviewer: ```suggestion
<img src="assets/53_image_search_datasets/spaces_image_search.jpg" alt="Screenshot of Gradio search app">
```

However, I'm a little bit vary about making this public straightaway. Looking at the model card for the CLIP model we can look at the primary intended uses:

> ### Primary intended uses
>

> > We primarily imagine the model will be used by researchers to better understand robustness, generalization, and other capabilities, biases,  and constraints of computer vision models.

Reviewer: ```suggestion
> We primarily imagine the model will be used by researchers to better understand robustness, generalization, and other capabilities, biases, and constraints of computer vision models.
```

suggests that 'deployment' is not a good idea. Whilst the results I got are interesting, I haven't played around with the model enough yet (and haven't done anything more systematic to evaluate its performance and biases) to be confident about 'deploying' it. Another additional consideration is the target dataset itself. The images are drawn from books covering a variety of subjects and time periods. There are plenty of books which represent colonial attitudes and as a result some of the images included may represent certain groups of people in a negative way. This could potentially be a bad combo with a tool which allows any arbitrary text input to be encoded as a prompt.

There may be ways around this issue but this will require a bit more thought.

## Conclusion

Although we don't have a nice demo to show for it, we've seen how we can use `datasets` to:

- load images into the new `Image` feature type
- 'save' our work using `push_to_hub` and use this to move data between machines/sessions

> - create a `faiss` index for images that we can use to retrieve images from a text (or image) input

Reviewer: ```suggestion
- create a `faiss` index for images that we can use to retrieve images from a text (or image) input.
```

---
title: "Image search with 🤗 datasets"
thumbnail: /blog/assets/spaces_image_search.jpg 
---

<h1> Image search with 🤗 datasets </h1> 

<div class="blog-metadata">

>     <small>Published March 8, 2022.</small>

Reviewer 3: Please update

different prompts:

- For some of these I'll choose prompts which are a broad 'category' i.e. 'a musical instrument' or 'an animal', others are specific i.e. 'a guitar'.

- Out of interest I also tried a boolean operator: "An illustration of a cat or a dog".

- Finally I tried something a little more abstract: \"an empty abyss\"

``` python
prompts = ["A musical instrument", "A guitar", "An animal", "An illustration of a cat or a dog", "an empty abyss"]
```

``` python
for prompt in prompts:
    get_image_from_text(prompt)
```

<img src="assets/53_image_search_datasets/musical_instrument.jpg">

<img src="assets/53_image_search_datasets/guitar.jpg">

<img src="assets/53_image_search_datasets/an_animal.jpg">

<img src="assets/53_image_search_datasets/cat_or_dog.jpg">

<img src="assets/53_image_search_datasets/an_empty_abyss.jpg">

We can see these results aren't always right but they are usually some reasonable results in there. It already seems like this could be useful for searching for a the semantic content of an image in this dataset. However we might hold off on sharing this as is...

## Creating a huggingface space? 🤷🏼 

> One obvious next step for this kind of project is to create a hugginface [spaces](https://huggingface.co/spaces) demo. This is what I've done for other [models](https://huggingface.co/spaces/BritishLibraryLabs/British-Library-books-genre-classifier-v2)

Reviewer 3: ```suggestion
One obvious next step for this kind of project is to create a Hugging Face [Spaces](https://huggingface.co/spaces/launch) demo. This is what I've done for other [models](https://huggingface.co/spaces/BritishLibraryLabs/British-Library-books-genre-classifier-v2)
```

```

We can now use the load `from_dict` method to create a new dataset.

``` python
from datasets import Dataset
```

``` python
dataset = Dataset.from_dict(data)
```

We can look at one example to see what this looks like.

``` python
dataset[0]
```

```
{'fname': '003557206_04_000013_1_The Works of the Reverend Dr  Jonathan Swift  Dean of St  Patrick s  Dublin  in twenty volumes  Containing_1772.jpg',
'path': 'embellishments_sample/1772/003557206_04_000013_1_The Works of the Reverend Dr  Jonathan Swift  Dean of St  Patrick s  Dublin  in twenty volumes  Containing_1772.jpg',
'year': '1772'}
```

### Loading our images

At the moment our dataset has the filename and full path for each image. However, we want to have an actual image loaded into our dataset. We can use the `cast_column` method to change the type of a column. Since we might want to keep our original `path` column, let's first make a copy of that column and store it in a new `img` column. We can do this by using `map` and returning a dictionary with the key `img` and the value from our `path` column.

``` python
dataset = dataset.map(lambda example: {"img": example['path']})
```

``` python

> dataset = dataset.cast_column('img', features.Image())

Author: > It would be amazing to showcase the new ImageFolder dataset (PR [here](https://github.com/huggingface/datasets/pull/2830/files)) in this blog post. This way, the entire pipeline, from loading to preprocessing, will be cached by default.

That sounds great, I've been watching that PR will great anticipation since it will be useful for another project I'm working on 😀

If that PR will be ready to merge soon I'm happy to update the blog to use that. I'll update the notebook/post tomorrow and ping you if I run into any problems. I assume that the public facing API won't change from what is outlined in the current PR?

> dataset = dataset.cast_column('img', features.Image())

Author: I have just taken a look at this. Unless I missed something in this case using `ImageFolder` might not work for all the parts in this example. In this dataset the structure follows the imagefolder approach for the most part i.e. 
```
top_folder
	year/
		fname.jpg
```

However, in this case the filename also contains some metadata.

It's possible I missed something but I think the fname isn't preserved when loaded using `ImageFolder`. This [Colab notebook](https://colab.research.google.com/drive/129zbECz8O1UgIm6pjGlcia46EWNya3iW?usp=sharing) shows what I have for this dataset. 

Do you think there is some benefit in preserving the original path as a feature in the dataset when loaded using `ImageFolder`? I think it's not super common to have metadata in the filename but it's also not unkown. This might also make it easier to handle folder which have the structure:

``` 
top_folder
	train/
		dogs/
			dog.1.jpg
		cats/
		pigeon/
	valid/
		dogs/
		cats/
		pigeon/
```

Since you would have access to the path `top_folder/train/dogs/dog1.jpg` and could use that information for splitting in the same way as in the original dataset i.e. some form of `map(lambda x: {"is_train": x['fpath'].split('/')[:-2]=="train"})

> dataset = dataset.cast_column('img', features.Image())

Author: I have just realised that you can get this information from the `Image.filename`. i.e. `dataset['train'][0]['image'].filename`.  I can use that for this example to get the other required metadata. I am still wondering if making this a feature would make sense (maybe as a flag) but there might be other downsides I haven't considered.

> dataset = dataset.cast_column('img', features.Image())

Reviewer 4: Yes, the public-facing API should stay the same.

Besides `image.filename`, another option is to use:
```python
dataset = dataset.cast_column("image", Image(decode=False))
dataset[0]["image"]["path"]
```
to access path information. My rationale for not having paths as a separate feature is to avoid storing duplicate data in a dataset, and additionally, it avoids local path information leakage in the `push_to_hub`

And to build splits, you can use globbing to match local files or files on the Hub, e.g.:
```python
from datasets import load_dataset
dataset = load_datataset("imagefolder", data_files={"train": "top_folder/train/**/*.jpg", "valid": "top_folder/valid/**/*.jpg"})
```

> dataset = dataset.cast_column('img', features.Image())

Author: Thanks, I've just updated to use `ImageFolder` for loading. Everything worked smoothly for this and it's really nice to be able to cut out a whole bunch of loading code using this 😀 For now, in the Colab I am installing from that branch but I will update that if this pull request is likely to make it into the next `datasets` release.

```python
dataset = dataset.filter(lambda example: example['img'] is not None)
```

Taking a look at our dataset again.

```python
dataset
```

```python
Dataset({
    features: ['fname', 'year', 'path', 'img'],
    num_rows: 416935
    })
```

You'll see we lost a few rows by doing this filtering. We should now just have images which are successfully loaded. 

If we access an example and index into the `img` column we'll see our image 😃

```python
dataset[10]['img']
```

## Push all the things to the hub! 

![Push all the things to the hub!](https://i.imgflip.com/613c0r.jpg)

One of the super awesome things about the huggingface ecosystem is the huggingface hub. We can use the hub to access models and datasets. Often this is used for sharing work with others but it can also be a useful tool for work in progress. The datasets library recently added a `push_to_hub` method that allows you to push a dataset to the hub with minimal fuss. This can be really helpful by allowing you to pass around a dataset with all the transformers etc. already done. 

> When I started playing around with this feature I was also keen to see if it could be used as a way of 'bundling' everything together. This is where I noticed that if you push a dataset containing images which have been loaded in from filepaths by pillow the version on the hub won't have the images attached. If you always have the image files in the same place when you work with the dataset then this doesn't matter. If you want to have the images stored in the parquet file(s) associated with the dataset we need to load it without the filename attribute present (there might be another way of ensuring that datasets doesn't rely on the image file being on the file system -- if you of this I'd love to hear about it). 

Author: Very excited about this feature, it's already been really useful for another project I'm working on.
